src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.cs

```diff
@@ -2,33 +2,74 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+
 namespace System.Threading
 {
     /// <summary>
     /// A LIFO semaphore.
     /// Waits on this semaphore are uninterruptible.
     /// </summary>
-    internal sealed class LowLevelLifoSemaphore : IDisposable
+    internal sealed partial class LowLevelLifoSemaphore : IDisposable
     {
-        private WaitSubsystem.WaitableObject _semaphore;
+        private WaiterListEntry _waiterStackHead;
+        private LowLevelLock _waiterStackLock;
+        [ThreadStatic]
+        private static WaiterListEntry t_waitEntry;
 
-        public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount)
+        private void Create(int maximumSignalCount)
         {
-            _semaphore = WaitSubsystem.WaitableObject.NewSemaphore(initialSignalCount, maximumSignalCount);
+            _waiterStackHead = null;
+            _waiterStackLock = new LowLevelLock();
         }
 
-        public bool Wait(int timeoutMs)
+        public void Dispose()
         {
-            return WaitSubsystem.Wait(_semaphore, timeoutMs, false, true);
         }
 
-        public int Release(int count)
+        private bool WaitForWake(int timeoutMs)
         {
-            return WaitSubsystem.ReleaseSemaphore(_semaphore, count);
+            WaiterListEntry waitEntry = t_waitEntry ?? (t_waitEntry = new WaiterListEntry());
+            waitEntry._monitor.Acquire();
+            try
+            {
+                _waiterStackLock.Acquire();
+                waitEntry._next = _waiterStackHead;
+                _waiterStackHead = waitEntry;
+                _waiterStackLock.Release();
+                return waitEntry._monitor.Wait(timeoutMs);
+            }
+            finally
+            {
+                waitEntry._monitor.Release();
+            }
         }
 
-        public void Dispose()
+        private void Wake(int count)
         {
+            while (count-- > 0)
+            {
+                _waiterStackLock.Acquire();
+                WaiterListEntry waitEntry = _waiterStackHead;
+                _waiterStackHead = waitEntry?._next;
+                _waiterStackLock.Release();
+                if (waitEntry != null)
+                {
+                    waitEntry._monitor.Acquire();
+                    waitEntry._monitor.Signal_Release();
+                }
+            }
+        }
+
+        class WaiterListEntry
+        {
+            public LowLevelMonitor _monitor;
+            public WaiterListEntry _next;
```

Comment:
I didn't split up `LowLevelLifoSemaphore` and `ClrLifoSemaphore` because technically there's only one user of it - the thread pool. There's probably never going to be another one because there's SemaphoreSlim, so as long as you specifically don't depend on uninterruptibility and LIFO order (at the same time) there's no reason to use this one. LowLevelLock already contained spinning code, so it seemed acceptable that spinning is part of the low-level code.

The runtime sharing is a bit of a PITA part that I had to put some thought into. I am trying hard to avoid a Mono-specific implementation (as long as I can avoid hurting CoreRT).

While the wait subsystem seems like an obvious choice on CoreRT it doesn't really save any code. In fact it makes the subsystem support LIFO order and uninterruptible Wait which is way more code than the primitive implementation here. Also the current implementation of WaitSubsystem has contention on a single lock, which hurts the performance a bit (implementation detail, but a real problem).

My current plan is to break this dependency on WaitSubsystem on this one place and move LowLevelLock/LowLevelMonitor/LowLevelSemaphore into shared part of CoreLib as a feature (eg. FeaturePortableLowLevelSync). That would be shared by CoreRT and Mono. CoreCLR could use it (for testing at least), or simply skip the feature and implement the primitives on its own through existing PAL code.

Reply:
Merging `LowLevelLifoSemaphore` and `ClrLifoSemaphore` is fine, the wait portions can still be specialized for runtimes, and I think they should be. I really don't want to use this implementation of a LIFO semaphore in CoreRT or CoreCLR, it's unnecessarily more complicated. The tracking being done here and additional overhead is unnecessary for those runtimes, which already do similar tracking to support all of the other types of waits and the tracking is shared so there is no additional overhead with uninterruptibility. Adding uninterruptibility in both runtimes was actually very easy. Conversely it was actually a lot more work to add interruptibility to CoreRT's wait subsystem, which started off uninterruptible during initial implementation. If it's not easy to modify Mono's semaphore to support uninterruptible waits, I would still prefer to separate out this implementation specifically for Mono.

> The `LowLevel` prefix also makes it immediately obvious that it is uninterruptible (as long as one knows the convention or read the long comment in WaitSubsystem).

At the moment the `LowLevel` part doesn't just mean uninterruptible, it also means it's a simple implementation that directly uses system components. `LowLevelLock` does some spinning, in hindsight it probably would have made sense to separate it out. The idea was that the `LowLevel` portions would be system dependencies that may be specific to runtimes (which may also be shared if possible) and anything on top of them could be shared.

> While the wait subsystem seems like an obvious choice on CoreRT it doesn't really save any code. In fact it makes the subsystem support LIFO order and uninterruptible Wait which is way more code than the primitive implementation here.

Not really, most of the code is shared with interruptible waits so there is no overhead that is specific to uninterruptible waits.

> Also the current implementation of WaitSubsystem has contention on a single lock which hurts the performance a bit (implementation detail, but a real problem).

I don't see how this implementation would be better. Typically these locks are uncontended. It's rare for two threads to both be releasing threads because the thread pool uses an interlocked operation to figure out how many threads need to be released and one of those two threads would release all of them. Perf is not a big concern but the wait subsystem is actually pretty fast.

Reply:
> Even if there is only one user currently, it can help to componentize the functionality.

Granted sometimes it can be advantageous to not componentize in order to add some features

Reply:
> I really don't want to use this implementation of a LIFO semaphore in CoreRT or CoreCLR, it's unnecessarily more complicated.

I do have different goals for this, which makes it hard agenda to push. I think it's not unnecessarily more complicated because the amount of code removed from the wait subsystem is more than the amount of code added here. I will do some more benchmarks of different variations, but I actually think that the contention on `s_lock` in `WaitSubsystem` hurts the performance significantly for heavily threaded code and this is an easy way out. The suggestion with CoreCLR was not about using it there for production code, but only as a means of bring up of the code in another runtime and for apples-to-apples performance comparison.

> At the moment the `LowLevel` part doesn't just mean uninterruptible, it also means it's a simple implementation that directly uses system components.

The documentation states

```
    /// <see cref="LowLevelLock"/> and <see cref="LowLevelMonitor"/>
    ///   - These are "low level" in the sense they don't depend on this wait subsystem, and any waits done are not
    ///     interruptible
```

LowLevelSemaphore never used the system component in Unix, it used the wait subsystem, violating any common sense about the component layering. In that sense calling it "low level" is wrong. The only other shared property is the uninterruptibility. I think it should be implemented using system components, but if I fail to push that direction then it should definitely be renamed to avoid confusion.

> Not really, most of the code is shared with interruptible waits so there is no overhead that is specific to uninterruptible waits.

No overhead in terms of performance for other code paths through the WaitSubsystem. Amount of code is different story though. Performance-wise it increases contention on the single lock. It's orthogonal issue to these code changes though.

Reply:
> I do have different goals for this, which makes it hard agenda to push.

Requirements for Mono should not need to change implementation in a different runtime unless it's actually better. It should be a simple matter of providing a small implementation (this implementation) on the Mono side. There will always be runtime-specific pieces and it's not a bad thing. It is also not necessary to port the wait subsystem, that is a runtime-specific piece currently. Perhaps it could be shared in the future but it doesn't have to be at the moment. Is there some other agenda you're trying to push for?

I'd like to know what is the issue with the implementation using the wait subsystem for CoreRT that this implementation would solve. As far as I'm aware, there was no issue and as such I don't see a need to change it. Design-wise, in CoreRT it totally makes sense to reuse the existing infrastructure instead of solving the problem again. The wait subsystem already does all of what this implementation does for the purpose of normal waits and the same thing is reused conveniently with no additional overhead. This implementation adds another monitor per thread, allocates another object per thread, has issues with OOMs that can be solved but makes it more complicated, and all of that is unnecessary in a runtime that has a clear alternative where those problems are already solved properly. I see many cons and the only pro is that the code is shareable, which to me is not a strong reason by itself. I don't see a strong reason to change CoreRT's implementation.

> I think it's not unnecessarily more complicated because the amount of code removed from the wait subsystem is more than the amount of code added here.

I don't see that, effectively the code removed is `RegisterPrioritizedWait`, a small and simple function, and the one call to it. The rest of it is just trivial plumbing and excluding the tests that are only relevant to interruptible waits.

> I will do some more benchmarks of different variations, but I actually think that the contention on `s_lock` in `WaitSubsystem` hurts the performance significantly for heavily threaded code and this is an easy way out.

The global lock is held for a very short time, I have not seen significant contention on that lock. If there is heavy contention using the wait subsystem that this implementation solves, I'd like to see an example.

> for apples-to-apples performance comparison

I believe it would be apples-to-apples regardless of which implementation is used for this type, perf-wise it wouldn't make much difference (except for a few things like allocation that may need fixing in this implementation). This implementation could be under a `#if MONO` or some other build-time define and that would allow comparing the same implementation of this type between runtimes if you really want to.

> LowLevelSemaphore never used the system component in Unix, it used the wait subsystem, violating any common sense about the component layering. In that sense calling it "low level" is wrong. The only other shared property is the uninterruptibility. I think it should be implemented using system components, but if I fail to push that direction then it should definitely be renamed to avoid confusion.

- I didn't find a system equivalent on Unixes
- The intention is different from the implementation, the fact that it uses the wait subsystem is an implementation detail
- That comment is specific to `LowLevelLock`, it intends to state that the implementation of this lock cannot depend on the wait subsystem (because it is used by the wait subsystem). It does not imply that "low level" means that it cannot use the wait subsystem, and that is not the case with `LowLevelLifoSemaphore`.
- I think the name is appropriate, I'm not too particular about the name as long as it sort of works, it's difficult to get names right.
- If a particular runtime does not want it to have a dependency on something then it can provide an implementation that does not need to have that dependency, that's not an issue. That desire for removing the dependency however, does not require removing a sensible dependency in a different runtime.

Reply:
re the naming of `LowLevelSemaphore`: I chose that name moreso to match the naming of `LowLevelLinq` (subset of API surface reimplemented for runtime use) than `LowLevelLock` or `LowLevelMonitor` (thin managed wrappers around native system components).

Reply:
> It is also not necessary to port the wait subsystem, that is a runtime-specific piece currently. Perhaps it could be shared in the future but it doesn't have to be at the moment. Is there some other agenda you're trying to push for?

The whole point is to share as much code as possible between the runtimes.

Code that is runtime specific doesn't get as much testing as the one that is shared. Likewise the performance improvements have historically been done mostly on the CoreCLR side and didn't translate to CoreRT which was a shame. Now that Mono is sharing the CoreLib code there was a push that helped to fix the gap by moving more of the code to the shared part of the CoreLib.

Ideally I would like to see the wait subsystem shared between all three runtimes, but we are still quite far away from there and it may well never happen.

Decoupling individual components - thread pool, timers, low level synchronization primitives - was an attempt to get there faster, piece by piece. Unfortunately I underestimated the complexity of the LIFO semaphore implementation and its edge cases. I assumed I could do it in less than 100 lines and simplify the WaitSubsystem at the same time to the point that it would be a win-win situation overall (no regression, cleaner component separation, the benefits of code being shared and tested by 2 runtimes and allowing testing the whole managed pool on CoreCLR as well). Eventually I ended up with an implementation that is small enough and covers all the edge cases that were brought up, but it requires unmanaged code and it's not as mature as whatever WaitSubsystem does now.

> This implementation adds another monitor per thread, allocates another object per thread, has issues with OOMs...

Yeah, it was flawed. I have completely rewritten it in the end, but it's out of scope of this PR. I solved couple of the issues by moving the LIFO semaphore to unmanaged code, reducing the number of PThread objects (1 mutex + 1 condition/waiting thread) and storing the LIFO entries on stack.

> The global lock is held for a very short time, I have not seen significant contention on that lock.

I have seen it in the heavy threading+locking benchmarks before, but it could have been flaw in the benchmark methodology. I went ahead and re-run a different combinations of the code with my benchmarks. It turns out that the additional logic with interlocked `Counts` loops avoids getting to the slow path in those heavy threading scenarios.

> I think the name is appropriate, I'm not too particular about the name as long as it sort of works, it's difficult to get names right.

I understand it is difficult. I am not happy about the name, but I don't really have a better name right now. The "low level" part makes me feel it should not call into something "higher level" like WaitSubsystem, implementation detail or not.

Reply:
> re the naming of `LowLevelSemaphore`: I chose that name moreso to match the naming of `LowLevelLinq` (subset of API surface reimplemented for runtime use) than `LowLevelLock` or `LowLevelMonitor` (thin managed wrappers around native system component

Ah, thanks for helpful historical note :-) `LowLevelLinq` was not really used in that part of code anymore, aside from a `using` in one of the files that seems like a leftover.

Reply:
A good time to share the implementation would be if a system LIFO wait support were to be found. This implementation of the component can be implemented in managed or native code, would be fairly small, and perf doesn't really matter, that I don't see a huge benefit in sharing it at the moment, especially since there are better alternatives for some runtimes. Sharing the portion above this part (the `ClrLifoSemaphore` part of tracking spinners/waiters, etc.) is a lot more useful.

Reply:
> A good time to share the implementation would be if a system LIFO wait support were to be found.

Not going to happen, at least not in a portable way. Every other library I found (notably Facebook Folly, which is referenced from couple of academic papers) uses an implementation similar to the one I came up with.